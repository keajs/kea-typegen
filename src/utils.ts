import * as ts from 'typescript'
import * as path from 'path'
import { cloneNode } from '@wessberg/ts-clone-node'
import { visitProgram } from './visit/visit'
import { parsedLogicToTypeString } from './print/print'
import { AppOptions, NameType, ParsedLogic } from './types'
import { NodeBuilderFlags } from 'typescript'

export function logicSourceToLogicType(logicSource: string, appOptions?: AppOptions) {
    const program = programFromSource(logicSource)
    const [parsedLogic] = visitProgram(program)
    return parsedLogicToTypeString(parsedLogic, appOptions)
}

export function sourceToSourceFile(sourceCode: string, filename: string = 'logic.ts') {
    return ts.createSourceFile(filename, sourceCode, ts.ScriptTarget.ES5, true)
}

export function programFromSource(sourceCode: string) {
    const options = {}
    const compilerHost = ts.createCompilerHost(options)
    compilerHost.getSourceFile = (fileName) => (fileName === 'logic.ts' ? sourceToSourceFile(sourceCode) : undefined)
    return ts.createProgram(['logic.ts'], options, compilerHost)
}

export function isKeaCall(node: ts.Node, checker: ts.TypeChecker) {
    if (!ts.isIdentifier(node)) {
        return false
    }

    if (!node.parent || !ts.isCallExpression(node.parent)) {
        return false
    }

    const symbol = checker.getSymbolAtLocation(node)
    if (!symbol || symbol.getName() !== 'kea') {
        return false
    }

    const input = node.parent.arguments[0]

    if (!input || !ts.isObjectLiteralExpression(input)) {
        return false
    }

    return true
}

export function getAndGatherTypeNodeForDefaultValue(
    defaultValue: ts.Node,
    checker: ts.TypeChecker,
    parsedLogic: ParsedLogic,
): ts.TypeNode {
    let typeNode
    if (defaultValue) {
        if (ts.isAsExpression(defaultValue)) {
            gatherImports(defaultValue.type, checker, parsedLogic)
            typeNode = cloneNode(defaultValue.type)
            if (ts.isParenthesizedTypeNode(typeNode)) {
                typeNode = typeNode.type
            }
        } else if (
            defaultValue?.kind === ts.SyntaxKind.TrueKeyword ||
            defaultValue?.kind === ts.SyntaxKind.FalseKeyword
        ) {
            typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword)
        } else if (ts.isStringLiteralLike(defaultValue)) {
            typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)
        } else if (ts.isNumericLiteral(defaultValue)) {
            typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)
        } else if (ts.isArrayLiteralExpression(defaultValue) && defaultValue.elements.length === 0) {
            typeNode = ts.createArrayTypeNode(ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword))
        } else {
            const foundTypeNode = checker.typeToTypeNode(checker.getTypeAtLocation(defaultValue), undefined, undefined)
            gatherImports(foundTypeNode, checker, parsedLogic)
            typeNode = cloneNode(foundTypeNode)
        }
    } else {
        typeNode = ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
    }
    return typeNode
}

export function getParameterDeclaration(param: ts.ParameterDeclaration) {
    return ts.createParameter(
        undefined,
        undefined,
        undefined,
        ts.createIdentifier(param.name.getText()),
        param.initializer || param.questionToken ? ts.createToken(ts.SyntaxKind.QuestionToken) : undefined,
        param.type || ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
        undefined,
    )
}

export const toSpaces = (key) => key.replace(/(?:^|\.?)([A-Z])/g, (x, y) => ' ' + y.toLowerCase()).replace(/^ /, '')

export function getActionTypeCreator(parsedLogic: ParsedLogic) {
    return function (actionName) {
        return `${toSpaces(actionName)} (${parsedLogic.pathString})`
    }
}

export function cleanDuplicateAnyNodes(reducers: NameType[]): NameType[] {
    let newReducers = {}

    for (const reducer of reducers) {
        // if (reducer.name === 'featureFlag') {
        //     debugger
        // }
        if (!newReducers[reducer.name] || !isAnyUnknown(reducer.typeNode)) {
            newReducers[reducer.name] = reducer
        }
    }

    return Object.values(newReducers)
}

export function extractImportedActions(
    actionObjects: ts.Expression | ts.ObjectLiteralExpression,
    checker: ts.TypeChecker,
    parsedLogic: ParsedLogic,
) {
    let extraActions = {}

    if (ts.isObjectLiteralExpression(actionObjects)) {
        // actionObjects =  { [githubLogic.actionTypes.setRepositories]: () => ... }
        for (const property of actionObjects.properties) {
            // property.name ==> [githubLogic.actionTypes.setRepositories]
            if (ts.isComputedPropertyName(property.name)) {
                let propertyExpression = property.name.expression

                if (ts.isPropertyAccessExpression(propertyExpression)) {
                    const { name, expression } = propertyExpression
                    const actionName = name.escapedText

                    const nameSymbol = checker.getSymbolAtLocation(property.name)
                    const actionType = nameSymbol.escapedName as string

                    if (ts.isPropertyAccessExpression(expression)) {
                        // expression.expression ==> githubLogic.actionTypes
                        // expression.name ==> setRepositories

                        const symbol = checker.getSymbolAtLocation(expression.expression)
                        const symbolType = checker.getTypeOfSymbolAtLocation(symbol, expression.expression)

                        const actionCreatorsProperty = symbolType
                            .getProperties()
                            .find((p) => p.escapedName === 'actionCreators')
                        const actionCreators = actionCreatorsProperty?.valueDeclaration

                        if (
                            actionCreators &&
                            ts.isPropertySignature(actionCreators) &&
                            ts.isTypeLiteralNode(actionCreators.type)
                        ) {
                            const actionCreator = actionCreators.type.members.find(
                                (m) => (m.name as ts.Identifier)?.escapedText === actionName,
                            )

                            if (
                                actionCreator &&
                                ts.isPropertySignature(actionCreator) &&
                                ts.isFunctionTypeNode(actionCreator.type)
                            ) {
                                const type = checker.getTypeFromTypeNode(actionCreator.type)
                                const signature = type.getCallSignatures()[0]
                                const sigReturnType = signature.getReturnType()
                                const sigReturnTypeNode = checker.typeToTypeNode(
                                    sigReturnType,
                                    undefined,
                                    NodeBuilderFlags.NoTruncation,
                                )
                                gatherImports(sigReturnTypeNode, checker, parsedLogic)

                                extraActions[actionType] = cloneNode(actionCreator.type) //payload
                            }
                        }
                    }
                }
            }
        }
    }
    return extraActions
}

export function getLogicPathString(appOptions: AppOptions, fileName: string) {
    let cwd = process.cwd()
    if (appOptions?.rootPath) {
        cwd = path.resolve(cwd, appOptions.rootPath)
    }
    const pathString = path
        .relative(cwd, fileName)
        .replace(/^.\//, '')
        .replace(/\.[jt]sx?$/, '')
        .replace(/\//g, '.')

    return pathString
}

export function getFilenamesForSymbol(symbol: ts.Symbol): string[] | undefined {
    return symbol?.declarations
        .map((d) => d.getSourceFile().fileName)
        .filter((str) => !str.includes('/node_modules/typescript/lib/lib'))
}

/** gathers onto parsedLogic the TypeReference nodes that are declared in a different sourceFile */
export function gatherImports(input: ts.Node, checker: ts.TypeChecker, parsedLogic: ParsedLogic) {
    if (!input) {
        return
    }
    function getImports(requestedNode) {
        let node = requestedNode
        if (ts.isTypeReferenceNode(node)) {
            const symbol = checker.getSymbolAtLocation(node.typeName) || (node.typeName as any).symbol
            if (symbol) {
                storeExtractedSymbol(symbol, checker, parsedLogic)
            }
        }
        ts.forEachChild(requestedNode, getImports)
    }
    getImports(input)
}

export function storeExtractedSymbol(symbol: ts.Symbol, checker: ts.TypeChecker, parsedLogic: ParsedLogic) {
    const declaration = symbol.getDeclarations()[0]

    if (ts.isImportSpecifier(declaration)) {
        const importFilename = getFilenameForImportSpecifier(declaration, checker)
        if (importFilename) {
            addTypeImport(parsedLogic, importFilename, declaration.getText())
        } else {
            parsedLogic.typeReferencesInLogicInput.add(declaration.getText())
        }
        return
    }

    const files = getFilenamesForSymbol(symbol)
    if (files.length === 1) {
        // same file, add to logicType<...>
        if (
            ts.isTypeAliasDeclaration(declaration) ||
            ts.isInterfaceDeclaration(declaration) ||
            ts.isEnumDeclaration(declaration)  ||
            ts.isClassDeclaration(declaration)
        ) {
            if (files[0] === parsedLogic.fileName) {
                parsedLogic.typeReferencesInLogicInput.add(declaration.name.getText())
            } else {
                // but is it exported?
                addTypeImport(parsedLogic, files[0], declaration.name.getText())
            }
        }
    }
}

export function getFilenameForImportSpecifier(
    declaration: ts.ImportSpecifier,
    checker: ts.TypeChecker,
): string | void {
    let importNode: ts.Node = declaration
    while (importNode && !ts.isImportDeclaration(importNode)) {
        importNode = importNode.parent
    }
    if (ts.isImportDeclaration(importNode)) {
        const moduleSymbol = checker.getSymbolAtLocation(importNode.moduleSpecifier)
        const otherSourceFile = moduleSymbol?.getDeclarations()[0].getSourceFile()
        if (otherSourceFile) {
            return otherSourceFile.fileName || importNode.moduleSpecifier.getText()
        }
    }
}

function addTypeImport(parsedLogic: ParsedLogic, file: string, typeName: string) {
    if (!parsedLogic.typeReferencesToImportFromFiles[file]) {
        parsedLogic.typeReferencesToImportFromFiles[file] = new Set()
    }
    parsedLogic.typeReferencesToImportFromFiles[file].add(typeName)
}

export function arrayContainsSet(array: string[], setToContain: Set<string>): boolean {
    const arraySet = new Set(array)
    for (const str of setToContain) {
        if (!arraySet.has(str)) {
            return false
        }
    }
    return true
}

export function unPromisify(node: ts.Node): ts.Node {
    if (ts.isTypeReferenceNode(node) && (node.typeName as any)?.escapedText === 'Promise') {
        return node.typeArguments?.[0]
    }
    return node
}

export function isAnyUnknown(node?: ts.Node): boolean {
    if (!node) {
        return true
    }
    const unPromised = unPromisify(node)
    return (
        !unPromised ||
        unPromised.kind === ts.SyntaxKind.AnyKeyword ||
        unPromised.kind === ts.SyntaxKind.UnknownKeyword ||
        (ts.isTypeLiteralNode(unPromised) && unPromised.members.length === 0)
    )
}
