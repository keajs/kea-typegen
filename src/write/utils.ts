import { ParsedLogic } from '../types'
import { visit, types, parse } from 'recast'

export const t = types.namedTypes
export const b = types.builders

import { parseSync } from '@babel/core'

export function getAst(filename: string, source: string): any {
    return parse(source, {
        parser: {
            parse: (source) =>
                parseSync(source, {
                    presets: ['@babel/preset-env', '@babel/preset-react', '@babel/preset-typescript'],
                    filename: filename,
                    parserOpts: {
                        tokens: true, // recast uses this
                    },
                }),
        },
    })
}

export function isKeaCall(path: any): boolean {
    const stmt = path.node
    return (
        t.Identifier.check(stmt.callee) &&
        stmt.callee.name === 'kea' &&
        stmt.arguments[0] &&
        (t.ObjectExpression.check(stmt.arguments[0]) || t.ArrayExpression.check(stmt.arguments[0])) &&
        path.parentPath &&
        t.VariableDeclarator.check(path.parentPath.value) &&
        t.Identifier.check(path.parentPath.value.id)
    )
}

export function visitAllKeaCalls(
    ast: any,
    parsedLogics: ParsedLogic[],
    filename: string,
    callback: (args: { parsedLogic: ParsedLogic; logicName: string; path: any }) => any,
): void {
    visit(ast, {
        visitCallExpression: function (path) {
            if (!isKeaCall(path)) {
                this.traverse(path)
                return
            }
            const logicName = path.parentPath.value.id?.name
            if (!logicName) {
                console.warn(
                    `❗ Can not visit logic in "${filename}:${JSON.stringify(
                        path.node.loc,
                    )}" because it's not stored as a variable.`,
                )
                return
            }
            const parsedLogic = parsedLogics.find((l) => l.logicName === logicName)
            if (!parsedLogic) {
                console.error(
                    `❗ While trying to visit a logic, could not find logicName "${logicName}" in the list of logicNames (${Object.keys(
                        parsedLogics.map((l) => l.logicName),
                    ).join(', ')}) in the file: ${filename}`,
                )
                return
            }
            callback.bind(this)({ logicName, parsedLogic, path })
            return false
        },
    })
}

export function assureImport(ast: any, importFrom: string, importName: string, localName: string, hasImport: boolean) {
    if (hasImport) {
        visit(ast, {
            visitImportDeclaration(path) {
                if (
                    path.value.source &&
                    t.StringLiteral.check(path.value.source) &&
                    path.value.source.value === importFrom
                ) {
                    path.value.specifiers.push(b.importSpecifier(b.identifier(importName), b.identifier(localName)))
                }
                return false
            },
        })
    } else {
        ast.program.body = [
            b.importDeclaration(
                [b.importSpecifier(b.identifier(importName), b.identifier(localName))],
                b.stringLiteral(importFrom),
            ),
            ...ast.program.body,
        ]
    }
}
