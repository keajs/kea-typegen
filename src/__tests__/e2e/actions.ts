import { logicSourceToLogicType } from '../../utils'

test('actions - with a function', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: () => ({
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            })
        })
    `
    expect(logicSourceToLogicType(logicSource)).toMatchSnapshot()
})

test('actions - as an object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: {
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toMatchSnapshot()
})

test('actions - as a function returning a object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: function () {
                return {
                    updateName: (name: string) => ({ name }),
                    updateOtherName: (otherName: string) => ({ otherName }),
                }
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toMatchSnapshot()
})

test('actions - with random values instead of functions', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: {
                updateName: (name?: string) => ({ name }),
                withDefaultValue: (name = "john") => ({ name }),
                withDefaultValueAndType: (name: string = "john") => ({ name }),
                withTrue: true,
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toMatchSnapshot()
})

test('actions - with generic', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: {
                updateName: <T extends string>(name?: T) => ({ name }),
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toMatchSnapshot()
})