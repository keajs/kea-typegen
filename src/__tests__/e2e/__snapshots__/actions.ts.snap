// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions - as a function returning a object 1`] = `
"export interface logicType extends Logic {
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: {
                name: string;
            };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: {
                otherName: string;
            };
        });
    };
    actionKeys: {
        \\"update name (logic)\\": \\"updateName\\";
        \\"update other name (logic)\\": \\"updateOtherName\\";
    };
    actionTypes: {
        updateName: \\"update name (logic)\\";
        updateOtherName: \\"update other name (logic)\\";
    };
    actions: {
        updateName: (name: string) => void;
        updateOtherName: (otherName: string) => void;
    };
    defaults: {};
    events: {};
    key: undefined;
    listeners: {};
    path: [
        \\"logic\\"
    ];
    pathString: \\"logic\\";
    props: Record<string, unknown>;
    reducer: (state: any, action: any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    sharedListeners: {};
    values: {};
    _isKea: true;
    _isKeaWithKey: false;
}"
`;

exports[`actions - as an object 1`] = `
"export interface logicType extends Logic {
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: {
                name: string;
            };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: {
                otherName: string;
            };
        });
    };
    actionKeys: {
        \\"update name (logic)\\": \\"updateName\\";
        \\"update other name (logic)\\": \\"updateOtherName\\";
    };
    actionTypes: {
        updateName: \\"update name (logic)\\";
        updateOtherName: \\"update other name (logic)\\";
    };
    actions: {
        updateName: (name: string) => void;
        updateOtherName: (otherName: string) => void;
    };
    defaults: {};
    events: {};
    key: undefined;
    listeners: {};
    path: [
        \\"logic\\"
    ];
    pathString: \\"logic\\";
    props: Record<string, unknown>;
    reducer: (state: any, action: any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    sharedListeners: {};
    values: {};
    _isKea: true;
    _isKeaWithKey: false;
}"
`;

exports[`actions - with a function 1`] = `
"export interface logicType extends Logic {
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: {
                name: string;
            };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: {
                otherName: string;
            };
        });
    };
    actionKeys: {
        \\"update name (logic)\\": \\"updateName\\";
        \\"update other name (logic)\\": \\"updateOtherName\\";
    };
    actionTypes: {
        updateName: \\"update name (logic)\\";
        updateOtherName: \\"update other name (logic)\\";
    };
    actions: {
        updateName: (name: string) => void;
        updateOtherName: (otherName: string) => void;
    };
    defaults: {};
    events: {};
    key: undefined;
    listeners: {};
    path: [
        \\"logic\\"
    ];
    pathString: \\"logic\\";
    props: Record<string, unknown>;
    reducer: (state: any, action: any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    sharedListeners: {};
    values: {};
    _isKea: true;
    _isKeaWithKey: false;
}"
`;

exports[`actions - with random values instead of functions 1`] = `
"export interface logicType extends Logic {
    actionCreators: {
        updateName: (name?: string) => ({
            type: \\"update name (logic)\\";
            payload: {
                name: string;
            };
        });
        withDefaultValue: (name?: any) => ({
            type: \\"with default value (logic)\\";
            payload: {
                name: string;
            };
        });
        withDefaultValueAndType: (name?: string) => ({
            type: \\"with default value and type (logic)\\";
            payload: {
                name: string;
            };
        });
        withTrue: () => ({
            type: \\"with true (logic)\\";
            payload: {
                value: true;
            };
        });
    };
    actionKeys: {
        \\"update name (logic)\\": \\"updateName\\";
        \\"with default value (logic)\\": \\"withDefaultValue\\";
        \\"with default value and type (logic)\\": \\"withDefaultValueAndType\\";
        \\"with true (logic)\\": \\"withTrue\\";
    };
    actionTypes: {
        updateName: \\"update name (logic)\\";
        withDefaultValue: \\"with default value (logic)\\";
        withDefaultValueAndType: \\"with default value and type (logic)\\";
        withTrue: \\"with true (logic)\\";
    };
    actions: {
        updateName: (name?: string) => void;
        withDefaultValue: (name?: any) => void;
        withDefaultValueAndType: (name?: string) => void;
        withTrue: () => void;
    };
    defaults: {};
    events: {};
    key: undefined;
    listeners: {};
    path: [
        \\"logic\\"
    ];
    pathString: \\"logic\\";
    props: Record<string, unknown>;
    reducer: (state: any, action: any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    sharedListeners: {};
    values: {};
    _isKea: true;
    _isKeaWithKey: false;
}"
`;

exports[`actions - with generic 1`] = `
"export interface logicType extends Logic {
    actionCreators: {
        updateName: <T extends string>(name?: T) => ({
            type: \\"update name (logic)\\";
            payload: {
                name: T;
            };
        });
    };
    actionKeys: {
        \\"update name (logic)\\": \\"updateName\\";
    };
    actionTypes: {
        updateName: \\"update name (logic)\\";
    };
    actions: {
        updateName: <T extends string>(name?: T) => void;
    };
    defaults: {};
    events: {};
    key: undefined;
    listeners: {};
    path: [
        \\"logic\\"
    ];
    pathString: \\"logic\\";
    props: Record<string, unknown>;
    reducer: (state: any, action: any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    sharedListeners: {};
    values: {};
    _isKea: true;
    _isKeaWithKey: false;
}"
`;
